optuna/samplers/_tpe/sampler.py

```diff
@@ -742,52 +734,34 @@ def _calculate_weights_below_for_multi_objective(
     below_trials: list[FrozenTrial],
     constraints_func: Callable[[FrozenTrial], Sequence[float]] | None,
 ) -> np.ndarray:
-    loss_vals = []
-    feasible_mask = np.ones(len(below_trials), dtype=bool)
-    for i, trial in enumerate(below_trials):
-        # Hypervolume contributions are calculated only using feasible trials.
-        if constraints_func is not None:
-            if any(constraint > 0 for constraint in constraints_func(trial)):
-                feasible_mask[i] = False
-                continue
-        values = []
-        for value, direction in zip(trial.values, study.directions):
-            if direction == StudyDirection.MINIMIZE:
-                values.append(value)
-            else:
-                values.append(-value)
-        loss_vals.append(values)
-    lvals = np.asarray(loss_vals, dtype=float)
-
-    # Calculate weights based on hypervolume contributions.
-    n_below = len(lvals)
-    weights_below: np.ndarray
-    if n_below == 0:
-        weights_below = np.asarray([])
-    elif n_below == 1:
-        weights_below = np.asarray([1.0])
+    if constraints_func is None:
+        is_feasible = np.ones(len(below_trials), dtype=bool)
     else:
-        worst_point = np.max(lvals, axis=0)
-        reference_point = np.maximum(1.1 * worst_point, 0.9 * worst_point)
-        reference_point[reference_point == 0] = EPS
-        hv = compute_hypervolume(lvals, reference_point)
-        indices_mat = ~np.eye(n_below).astype(bool)
-        contributions = np.asarray(
-            [
-                hv - compute_hypervolume(lvals[indices_mat[i]], reference_point)
-                for i in range(n_below)
-            ]
-        )
-        contributions[np.isnan(contributions)] = np.inf
-        max_contribution = np.maximum(np.max(contributions), EPS)
-        if not np.isfinite(max_contribution):
```

Comment:
From the monotonicity of the hypervolume (submodularity), `hv` is always larger than the leave-one-out hypervolume, i.e. `compute_hypervolume(lvals[indices_mat[i]], reference_point)`, meaning that each element of `contributions` becomes `inf` only if `hv` is `inf`.

Since we also define the line `contributions[np.isnan(contributions)] = np.inf`, which takes care of `inf - inf = inf` in the hypervolume contribution calculation, if any element in `contributions` takes `inf`, all the elements in `contributions` become `inf`.

That is why we can separate this exception handling as the processing only for `np.isinf(hv) == True`.